odoo/addons/base/models/res_users.py

```diff
@@ -1124,20 +1124,30 @@ def create(self, vals_list):
                 values['groups_id'] = type(self).groups_id.convert_to_write(gs, user)
         return super(UsersImplied, self).create(vals_list)
 
+    def _clear_demoted_user_access(self, values, users_before):
+        portal_users_to_clear = self.env['res.users']
+        for user in self:
+            if not user.has_group('base.group_user') and user in users_before:
+                portal_users_to_clear |= user
```

Comment:
By construction, the following set relations holds: `portal_users_to_clear <= users_before <= self`.
The loop (with O(n²) time complexity, by the way) is equivalent to
```suggestion
        portal_users_to_clear = users_before.filtered(lambda u: not u.has_group('base.group_user'))
```
And by the way, the adjective "portal" is more specific than it should.  I suggest naming that recordset `demoted_users` instead.